File: src/main/java/bad/robot/radiate/teamcity/AllProjectsMonitor.java
package bad.robot.radiate.teamcity;

import bad.robot.http.HttpClient;
import bad.robot.radiate.Aggregator;
import bad.robot.radiate.monitor.Information;
import bad.robot.radiate.monitor.MonitoringTask;
import bad.robot.radiate.monitor.NonRepeatingObservable;
import com.googlecode.totallylazy.Sequence;

import static bad.robot.radiate.Aggregator.aggregate;
import static bad.robot.radiate.Functions.asString;
import static com.googlecode.totallylazy.Sequences.sequence;
import static java.lang.String.format;

/** @see {@link bad.robot.radiate.MonitoringTypes#singleAggregate()} */
public class AllProjectsMonitor extends NonRepeatingObservable implements MonitoringTask {

    private final TeamCityConfiguration configuration;
    private final HttpClient http;
    private final Server server;
    private final TeamCity teamcity;

    private Sequence<String> monitored = sequence("unknown");

    public AllProjectsMonitor(TeamCityConfiguration configuration) {
        this.configuration = configuration;
        this.server = new Server(configuration.host(), configuration.port());
        this.http = new HttpClientFactory().create(configuration);
        this.teamcity = new TeamCity(server, configuration.authorisation(), http, new JsonProjectsUnmarshaller(), new JsonProjectUnmarshaller(), new JsonBuildUnmarshaller());
    }

    @Override
    public void run() {
        try {
            Iterable<Project> projects = configuration.filter(teamcity.retrieveProjects());
            monitored = sequence(projects).map(asString());
            Iterable<BuildType> buildTypes = teamcity.retrieveBuildTypes(projects);
            Sequence<Build> builds = sequence(buildTypes).mapConcurrently(buildType -> teamcity.retrieveLatestBuild(buildType));
            Aggregator aggregated = aggregate(builds);
            notifyObservers(aggregated.activity(), aggregated.progress());
            notifyObservers(aggregated.status());
            notifyObservers(new Information(toString()));
        } catch (Exception e) {
            notifyObservers(e);
        }
    }

    @Override
    public String toString() {
        return format("monitoring %s as a single aggregate", monitored.toString("\r\n"));
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/AllProjectsMonitor.scala
package bad.robot.radiate.teamcity

import bad.robot.radiate.Aggregate.aggregate
import bad.robot.radiate.config.KnobsConfig
import bad.robot.radiate.config.Config
import bad.robot.radiate.monitor.{Information, MonitoringTask, NonRepeatingObservable}
import KnobsConfig._
import bad.robot.radiate.teamcity.ListBuildTypesSyntax._

import scalaz.syntax.either._

class AllProjectsMonitor extends NonRepeatingObservable with MonitoringTask {

  private var monitored = List("unknown")

  def run() {
    val builds = for {
      file        <- load() orElse create
      config      <- Config(file)
      _           <- store(file)
      http         = HttpClientFactory().create(config)
      teamcity     = TeamCity(TeamCityUrl(config.url), config.authorisation, http, new JsonProjectsUnmarshaller, new JsonProjectUnmarshaller, new JsonBuildUnmarshaller)
      all         <- teamcity.retrieveProjects
      projects    <- all.filter(project => config.projects.contains(project.id)).right
      monitored   <- projects.map(_.toString).toList.right
      buildTypes  <- teamcity.retrieveBuildTypes(projects)
      builds      <- buildTypes.getBuilds(teamcity)
      aggregation <- aggregate(builds).right
    } yield (monitored, aggregation)

    builds.fold(errors => {
      notifyObservers(errors)
    }, {
      case (monitored, aggregated) => {
        this.monitored = monitored
        notifyObservers(aggregated.activity, aggregated.progress)
        notifyObservers(aggregated.status)
        notifyObservers(new Information(toString))
      }
    })
  }

  override def toString = s"monitoring ${monitored.mkString("\r\n")} as a single aggregate"
}